### PR TITLE
[INDY-2129][INDY-2057][INDY-2077][INDY-2087] New tests and fixes

### PR DESCRIPTION
**Pull request by andkononykhin**
- adds systems tests for
  - INDY-1229: primaries updates after primary demotion
  - INDY-2077: forbidden constraints
  - INDY-2087: auth ruleS txn
- improves test fixtures and helpers
- fixes some tests
- improves docker routine and docs

### CODE REVIEW

File: system/utils.py [hunk 1]
```diff
@@ -79,12 +85,63 @@ def random_seed_and_json():
...
+                sleep_dur = retry_wait
+                if verbose:
+                    logger.debug("{} not succeeded yet, {:.2f} seconds "
+                                 "remaining..., will sleep for {}".format(fname, remain, sleep_dur))
+                await asyncio.sleep(sleep_dur)
+            else:
+                logger.error("{} failed; not trying any more because {} "
+                             "seconds have passed; args were {}".
+                             format(fname, timeout, args))
+                raise ex
```
Comment (skhoroshavin) [inline system/utils.py]: It would be nice to log exception as well - sometimes tracebacks get eaten by pytest

File: system/utils.py [hunk 2]
```diff
@@ -205,28 +262,99 @@ def wrapped(operations, queue_size, add_size, get_size, event_loop):
...
     add = await write_eventually_positive(send_nym, pool_handle, wallet_handle, trustee_did, some_did)
     assert add['op'] == 'REPLY'
     get = await read_eventually_positive(get_nym, pool_handle, wallet_handle, trustee_did, some_did)
     assert get['result']['seqNo'] is not None
 
 
-async def check_ledger_sync():
-    hosts = [testinfra.get_host('ssh://node{}'.format(i)) for i in range(1, 8)]
+# TODO make that async
+def check_no_failures(hosts):
```
Comment (skhoroshavin) [inline system/utils.py]: Why do you think this should be async?

File: system/utils.py [hunk 3]
```diff
@@ -205,28 +262,99 @@ def wrapped(operations, queue_size, add_size, get_size, event_loop):
...
+
+# TODO make that async
+async def check_ledger_sync(node_ids=None, nodes_num=7):
+    if node_ids is None:
+        node_ids = [(i + 1) for i in range(nodes_num)]
+    hosts = [NodeHost(i) for i in node_ids]
+
     pool_results = [host.run('read_ledger --type=pool --count') for host in hosts]
-    print('\nPOOL LEDGER SYNC: {}'.format([result.stdout for result in pool_results]))
+    print('\nPOOL LEDGER SYNC: {}'.format([result for result in pool_results]))
```
Comment (skhoroshavin) [inline system/utils.py]: Original output was more readable, however I do understand your concerns that stderr and rc were missing. What about creating some helper function that always outputs stdout, outputs stderr if it is not empty and return code if it is non zero?

File: system/utils.py [hunk 4]
```diff
@@ -205,28 +262,99 @@ def wrapped(operations, queue_size, add_size, get_size, event_loop):
...
-    print('\nCONFIG LEDGER SYNC: {}'.format([result.stdout for result in config_results]))
+    print('\nCONFIG LEDGER SYNC: {}'.format([result for result in config_results]))
     domain_results = [host.run('read_ledger --type=domain --count') for host in hosts]
-    print('\nDOMAIN LEDGER SYNC: {}'.format([result.stdout for result in domain_results]))
+    print('\nDOMAIN LEDGER SYNC: {}'.format([result for result in domain_results]))
     audit_results = [host.run('read_ledger --type=audit --count') for host in hosts]
-    print('\nAUDIT LEDGER SYNC: {}'.format([result.stdout for result in audit_results]))
+    print('\nAUDIT LEDGER SYNC: {}'.format([result for result in audit_results]))
+
+    assert all([pool_results[i] == pool_results[i + 1] for i in range(-1, len(pool_results) - 1)])
```
Comment (skhoroshavin) [inline system/utils.py]: More idiomatic way would be
```python
assert all(r1 == r2 for r1, r2 in zip(pool_results, pool_results[1:]))
```

File: system/utils.py [hunk 5]
```diff
@@ -205,28 +262,99 @@ def wrapped(operations, queue_size, add_size, get_size, event_loop):
...
+    assert all([config_results[i] == config_results[i + 1] for i in range(-1, len(config_results) - 1)])
+    assert all([domain_results[i] == domain_results[i + 1] for i in range(-1, len(domain_results) - 1)])
+    assert all([audit_results[i] == audit_results[i + 1] for i in range(-1, len(audit_results) - 1)])
+
+
+async def check_pool_performs_write_read(pool_handle, wallet_handle, trustee_did, nyms_count=1):
+    for i in range(nyms_count):
```
Comment (skhoroshavin) [inline system/utils.py]: Since you don't use index you can just write
```python
for _ in range(nyms_count):
```

File: system/utils.py [hunk 6]
```diff
@@ -205,28 +262,99 @@ def wrapped(operations, queue_size, add_size, get_size, event_loop):
...
+        some_did, _ = await did.create_and_store_my_did(wallet_handle, '{}')
+        add = await send_nym(pool_handle, wallet_handle, trustee_did, some_did)
+        assert add['op'] == 'REPLY'
+        get = await get_nym(pool_handle, wallet_handle, trustee_did, some_did)
+        assert get['result']['seqNo'] is not None
+
+
+async def check_pool_is_workable(pool_handle, wallet_handle, trustee_did, nyms_count=3):
```
Comment (skhoroshavin) [inline system/utils.py]: I'd prefer `check_pool_is_functional`, we have similar helper function in plenum with similar name.

File: system/utils.py [hunk 7]
```diff
@@ -205,28 +262,99 @@ def wrapped(operations, queue_size, add_size, get_size, event_loop):
...
+    await check_pool_performs_write_read(
+        pool_handle, wallet_handle, trustee_did, nyms_count=nyms_count
+    )
+
+
+async def ensure_pool_is_workable(pool_handle, wallet_handle, trustee_did, nyms_count=3, timeout=30):
```
Comment (skhoroshavin) [inline system/utils.py]: `ensure_pool_is_functional`?

File: system/utils.py [hunk 8]
```diff
@@ -205,28 +262,99 @@ def wrapped(operations, queue_size, add_size, get_size, event_loop):
...
+    assert all([audit_results[i] == audit_results[i + 1] for i in range(-1, len(audit_results) - 1)])
+
+
+async def check_pool_performs_write_read(pool_handle, wallet_handle, trustee_did, nyms_count=1):
+    for i in range(nyms_count):
+        some_did, _ = await did.create_and_store_my_did(wallet_handle, '{}')
+        add = await send_nym(pool_handle, wallet_handle, trustee_did, some_did)
+        assert add['op'] == 'REPLY'
+        get = await get_nym(pool_handle, wallet_handle, trustee_did, some_did)
+        assert get['result']['seqNo'] is not None
```
Comment (skhoroshavin) [inline system/utils.py]: Probably this can be replaced with `send_and_get_nym` helper?

File: system/utils.py [hunk 9]
```diff
@@ -205,28 +262,99 @@ def wrapped(operations, queue_size, add_size, get_size, event_loop):
...
+
+async def ensure_pool_is_workable(pool_handle, wallet_handle, trustee_did, nyms_count=3, timeout=30):
+    await eventually(
+        check_pool_is_workable, pool_handle, wallet_handle, trustee_did, nyms_count=nyms_count,
+        retry_wait=1, timeout=timeout
+    )
+
+
+async def check_pool_is_in_sync(node_ids=None, nodes_num=7):
+    await check_ledger_sync(node_ids=node_ids, nodes_num=nodes_num)
```
Comment (skhoroshavin) [inline system/utils.py]: Why do we need two functions with different names but same functionality?

File: system/utils.py [hunk 10]
```diff
@@ -205,28 +262,99 @@ def wrapped(operations, queue_size, add_size, get_size, event_loop):
...
+
+
+async def ensure_primary_changed(pool_handler, wallet_handler, trustee_did, primary_before):
+    return await eventually(
+        check_primary_changed, pool_handler, wallet_handler, trustee_did, primary_before,
+        retry_wait=1, timeout=30
+    )
+
+
+# TODO use threads to make that concurrent/async
```
Comment (skhoroshavin) [inline system/utils.py]: :+1:

File: system/utils.py [hunk 11]
```diff
@@ -248,7 +376,7 @@ def wrapped(operations, queue_size, add_size, get_size, event_loop):
                                                                                                   -len(':0')]
     except TypeError:
         try:
-            time.sleep(120)
+            await asyncio.sleep(120)
```
Comment (skhoroshavin) [inline system/utils.py]: I do understand that it was here originally, but:
- it looks like this whole thing can be replaced with `eventually`
- I suspect retry timeout can be reduced to 60-70 seconds

File: system/utils.py [hunk 12]
```diff
@@ -593,14 +701,14 @@ def get_pool_info(primary: str) -> dict:
     assert promote_res['op'] == 'REPLY'
 
 
-async def eventually_positive(func, *args, cycles_limit=15):
+async def eventually_positive(func, *args, cycles_limit=15, sleep=30, **kwargs):
```
Comment (skhoroshavin) [inline system/utils.py]: What about using same parameters (`retryTimeout` and `timeout`) as in `eventually`?

File: system/draft/TestAuditSuite.py [hunk 13]
```diff
@@ -84,7 +102,7 @@ class TestAuditSuite:
         p1.start_service()
         for node in test_nodes[5:]:
             node.restart_service()
-        time.sleep(30)
+        await asyncio.sleep(30)
```
Comment (skhoroshavin) [inline system/draft/TestAuditSuite.py]: Is sleep really needed here? If it is - can it be replaced by some `evenutal` check?

File: system/conftest.py [hunk 14]
```diff
@@ -37,10 +51,22 @@ def event_loop():
...
+@pytest.fixture(scope="session")
+def payment_init_session(request):
+    if request.config.getoption("payments"):
+        loop = asyncio.get_event_loop()
+        loop.run_until_complete(payment_initializer('libsovtoken.so', 'sovtoken_init'))
+
+
+@pytest.fixture
+def payment_init(request, payment_init_session):
+    aaa = request.config.getoption("payments", skip=True)
```
[the remark this answers is not in the document]
Comment (andkononykhin) [reply system/conftest.py]: to skip payment tests conditionally

File: system/conftest.py [hunk 15]
```diff
@@ -62,7 +88,6 @@ def event_loop():
 @pytest.fixture()
 @async_generator
 async def initial_fees_setting(pool_handler, wallet_handler, get_default_trustee):
-    await payment_initializer('libsovtoken.so', 'sovtoken_init')
```
[the remark this answers is not in the document]
Comment (andkononykhin) [reply system/conftest.py]: no, will fix, thanks

File: system/utils.py (hunk 4 above)
Comment (andkononykhin) [reply system/utils.py]: ok

File: system/utils.py (hunk 5 above)
Comment (andkononykhin) [reply system/utils.py]: agreed

File: system/utils.py (hunk 9 above)
Comment (andkononykhin) [reply system/utils.py]: that was fixed

File: system/conftest.py [hunk 16]
```diff
@@ -90,7 +115,58 @@ def event_loop():
...
+        return
+
+    config_entry = (
+        "Host node{}\n"
+        "\tHostName 10.0.0.{}\n"
+        "\tUser root\n"
+        "\tIdentityFile ~/.ssh/test_key\n"
+        "\tStrictHostKeyChecking no"
+    )
+    config = '\n'.join([config_entry.format(i + 1, i + 2) for i in range(nodes_num)])
```
[the remark this answers is not in the document]
Comment (andkononykhin) [reply system/conftest.py]: agreed, updated

File: system/conftest.py [hunk 17]
```diff
@@ -90,7 +115,58 @@ def event_loop():
...
+async def ssh_config(nodes_num):
+    if os.environ.get('IN_DOCKER_ENV') != 'yes':
+        return
+
+    config_entry = (
+        "Host node{}\n"
+        "\tHostName 10.0.0.{}\n"
+        "\tUser root\n"
+        "\tIdentityFile ~/.ssh/test_key\n"
+        "\tStrictHostKeyChecking no"
```
[the remark this answers is not in the document]
Comment (andkononykhin) [reply system/conftest.py]: sure, updated

File: system/conftest.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (andkononykhin) [reply system/conftest.py]: fixed

File: system/conftest.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (andkononykhin) [reply system/conftest.py]: added comments

**Pull request merged**